joint_trajectory_controller/test/test_trajectory_controller.cpp

```diff
@@ -441,3 +441,121 @@ TEST_F(TestTrajectoryController, correct_initialization_using_parameters) {
   ASSERT_EQ(State::PRIMARY_STATE_INACTIVE, state.id());
   executor.cancel();
 }
+
+
+TEST_F(TestTrajectoryController, test_state_publish_rate) {
+  auto traj_controller = std::make_shared<joint_trajectory_controller::JointTrajectoryController>();
+  auto ret = traj_controller->init(test_robot, controller_name);
+  if (ret != controller_interface::CONTROLLER_INTERFACE_RET_SUCCESS) {
+    FAIL();
+  }
+
+  // fill in some data so we wont fail
+  auto traj_lifecycle_node = traj_controller->get_lifecycle_node();
+  std::vector<std::string> joint_names = {"joint1", "joint2", "joint3"};
+  rclcpp::Parameter joint_parameters("joints", joint_names);
+  traj_lifecycle_node->set_parameter(joint_parameters);
+
+  std::vector<std::string> operation_mode_names = {"write1", "write2"};
+  rclcpp::Parameter operation_mode_parameters("write_op_modes", operation_mode_names);
+  traj_lifecycle_node->set_parameter(operation_mode_parameters);
+
+  rclcpp::executors::SingleThreadedExecutor executor;
+  executor.add_node(traj_lifecycle_node->get_node_base_interface());
+
+  traj_lifecycle_node->set_parameter(rclcpp::Parameter("state_publish_rate", 10));
+
+  traj_controller->on_configure(traj_lifecycle_node->get_current_state());
+  traj_controller->on_activate(traj_lifecycle_node->get_current_state());
+
+  auto future_handle = std::async(
+    std::launch::async, [&executor]() -> void {
+      executor.spin();
+    });
+
+  typedef joint_trajectory_controller::JointTrajectoryController::ControllerStateMsg Msg;
+
+  int echo_received_counter = 0;
+  rclcpp::Subscription<Msg>::SharedPtr subs = traj_lifecycle_node->create_subscription<Msg>(
+    "/state",
+    10,
+    [&](Msg::UniquePtr msg) {
+      (void)msg;
+      ++echo_received_counter;
+      // RCLCPP_INFO(rclcpp::get_logger("test_state_publisher"), "%d", echo_received_counter);
+    }
+  );
+
+  // wait for things to setup
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
Why wait for this amount? My concern is that `sleep_for`s will lead to flakey tests.
Is there a better way to check that the resources are ready?

Reply:
Not that I know of.

Other tests are using it as well, see [this](https://github.com/ros-controls/ros2_controllers/blob/master/joint_trajectory_controller/test/test_trajectory_controller.cpp#L330). We should do the fix (if we find it) in a single pass.

Reply:
Indeed. We had a lot of issues in `ros_controllers` with flaky tests, best to root them out early on!